File: frontend/src/modules/capacity/services/capacity.ts
import api from '@/core/services/client';
import type {
  PeriodInsight,
  PeriodProjectInsight,
  PeriodUserInsight,
  ReportableUser,
} from '@/modules/capacity/types/capacity';
import type { AllocationUsersResponse } from '@/modules/capacity/types/allocation';

export const capacityApi = {
  getInsights: async (startDate: string, endDate: string): Promise<PeriodInsight[]> => {
    const response = await api.get<PeriodInsight[]>('/capacity/insights', {
      params: { start_date: startDate, end_date: endDate },
    });
    return response.data;
  },
  getInsightsDetail: async (
    fa: string,
    startDate: string,
    endDate: string,
  ): Promise<PeriodUserInsight[]> => {
    const response = await api.get<PeriodUserInsight[]>('/capacity/insights/detail', {
      params: { fa, start_date: startDate, end_date: endDate },
    });
    return response.data;
  },
  getUserDetail: async (
    userId: string,
    startDate: string,
    endDate: string,
  ): Promise<PeriodProjectInsight[]> => {
    const response = await api.get<PeriodProjectInsight[]>('/capacity/insights/user-detail', {
      params: { user_id: userId, start_date: startDate, end_date: endDate },
    });
    return response.data;
  },
  getReportableUsers: async (): Promise<ReportableUser[]> => {
    const response = await api.get<ReportableUser[]>('/capacity/insights/user-detail/users');
    return response.data;
  },
  getAllocationUsers: async (): Promise<AllocationUsersResponse> => {
    const response = await api.get<AllocationUsersResponse>('/capacity/allocation/users');
    return response.data;
  },
};

File: mcp_server/data/capacity.py
"""Capacity data access — wraps core/services/capacity_insights.py for MCP."""

from __future__ import annotations

from datetime import date

from sqlalchemy.ext.asyncio import AsyncSession

from app.core.services.capacity_insights import (
    get_allocation_projects,
    get_allocation_users,
    get_capacity_fa_detail,
    get_capacity_insights,
    get_capacity_user_detail,
    get_reportable_users,
)


def _default_range() -> tuple[date, date]:
    """Default 6-month range ending at current month."""
    today = date.today()
    end = date(today.year, today.month, 1)
    start_month = today.month - 5
    start_year = today.year
    if start_month <= 0:
        start_month += 12
        start_year -= 1
    start = date(start_year, start_month, 1)
    return start, end


async def get_insights(
    session: AsyncSession,
    start_date: date | None = None,
    end_date: date | None = None,
) -> list[dict]:
    """FA-level billable allocation overview by period."""
    start, end = start_date, end_date
    if start is None or end is None:
        start, end = _default_range()
    return await get_capacity_insights(session, start, end)


async def get_fa_detail(
    session: AsyncSession,
    fa: str,
    start_date: date | None = None,
    end_date: date | None = None,
) -> list[dict]:
    """Per-user breakdown for a functional area."""
    start, end = start_date, end_date
    if start is None or end is None:
        start, end = _default_range()
    return await get_capacity_fa_detail(session, fa, start, end)


async def get_user_detail(
    session: AsyncSession,
    user_id: str,
    start_date: date | None = None,
    end_date: date | None = None,
) -> list[dict]:
    """Per-project breakdown for a user."""
    start, end = start_date, end_date
    if start is None or end is None:
        start, end = _default_range()
    return await get_capacity_user_detail(session, user_id, start, end)


async def get_allocation(
    session: AsyncSession,
    view: str = "users",
    start_date: date | None = None,
    end_date: date | None = None,
) -> dict:
    """Averaged allocation by users or projects over finished periods."""
    if view == "projects":
        return await get_allocation_projects(session, start_date, end_date)
    return await get_allocation_users(session, start_date, end_date)


async def get_users(session: AsyncSession) -> list[dict]:
    """List reportable users for selectors."""
    return await get_reportable_users(session)
